# Assignment: Build a Regression Model in Keras
## <a href="#parta"> Part-A click here</a>

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

# Part (B)-- <a href="#partb"> click here</a>

B. Normalize the data (5 marks)

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?

# Part (C)-- <a href='#partc'>click here </a>

C. Increate the number of epochs (5 marks)

Repeat Part B but use 100 epochs this time for training.

How does the mean of the mean squared errors compare to that from Step B?

In [2]:
# @title Importing Library
import pandas as pd
import numpy as np

# Library for modal
import keras
from keras.models import Sequential
from keras.layers import Dense

#For data spliting
from sklearn.model_selection import train_test_split

#For mean sqare error
from sklearn.metrics import mean_squared_error

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:521: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

In [3]:
# @title Loading data set-

concrete_data=pd.read_csv('https://cocl.us/concrete_data')

In [4]:
concrete_data.sample(5)

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
684    192.0               288.0      0.0  192.0               0.0   
469    194.7                 0.0    100.5  170.2               7.5   
429    213.5                 0.0    174.2  159.2              11.7   
699    203.5               305.3      0.0  203.5               0.0   
1008   160.2               188.0    146.4  203.2              11.3   

      Coarse Aggregate  Fine Aggregate  Age  Strength  
684              929.8           716.1   90     50.73  
469              998.0           901.8  100     44.28  
429             1043.6           771.9   14     29.59  
699              963.4           630.0    3      9.56  
1008             828.7           709.7   28     35.31

**The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:**

1. Cement

2. Blast Furnace Slag

3. Fly Ash

4. Water

5. Superplasticizer

6. Coarse Aggregate

7. Fine Aggregate

### Let's check how many data points we have.

In [5]:
concrete_data.shape

(1030, 9)

In [6]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [7]:
concrete_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Cement              1030 non-null   float64
 1   Blast Furnace Slag  1030 non-null   float64
 2   Fly Ash             1030 non-null   float64
 3   Water               1030 non-null   float64
 4   Superplasticizer    1030 non-null   float64
 5   Coarse Aggregate    1030 non-null   float64
 6   Fine Aggregate      1030 non-null   float64
 7   Age                 1030 non-null   int64  
 8   Strength            1030 non-null   float64
dtypes: float64(8), int64(1)
memory usage: 72.5 KB


In [8]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

There are **1030** samples in the dataset.

**Strength is our target**

**The data looks very clean and is ready to be used to build our model.**

In [9]:
# Split data into predictors and target

predictors = concrete_data.iloc[:,:-1] # strength is the last column so this will exclude the last column.
target = concrete_data['Strength'] # Strength column

In [10]:
predictors.sample(5)

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
658   305.3               203.5      0.0  203.5               0.0   
312   295.7                 0.0     95.6  171.5               8.9   
280   251.4                 0.0    118.3  188.5               6.4   
535   393.0                 0.0      0.0  192.0               0.0   
677   102.0               153.0      0.0  192.0               0.0   

     Coarse Aggregate  Fine Aggregate  Age  
658             965.4           631.0    7  
312             955.1           859.2   56  
280            1028.4           757.7   14  
535             940.6           785.6   90  
677             887.0           942.0    7

In [11]:
target.head(3)

0    79.99
1    61.89
2    40.27
Name: Strength, dtype: float64

In [12]:
# No. of features
n_cols=predictors.shape[1]
n_cols

8

The below function creates a model that has one hidden layer with 10 neurons and a ReLU activation function. It uses the adam optimizer and the mean squared error as the loss function.

Function is using keras Sequantial that we have imported above

# <div id='parta'>Part-(A)</div>

In [13]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,))) # hidden layers with node 10 and relu-- activation function
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [14]:
# Let's split the data in training and testing
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

### Train and Test the Network

**Let's call the function now to create our model.**

In [ ]:
# build the model
# model = regression_model()
# epochs=50

# # Fit
# model.fit(X_train, y_train, epochs=epochs, verbose=1)

In [16]:
# Evaluate the model on the test data.

loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

309/309 [==============================] - 0s 327us/step


204.88458617373962

### Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [18]:
# Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.
# Making it as a function as we will use it later--
def iterate(epochs,total_mean_squared_errors):
#     total_mean_squared_errors = 50
#     epochs = 50
    mean_squared_errors = []
    model = regression_model()
    # Iterating 50 times
    for i in range(0, total_mean_squared_errors):
        X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
        model.fit(X_train, y_train, epochs=epochs, verbose=0)
        MSE = model.evaluate(X_test, y_test, verbose=0)
        print("MSE "+str(i+1)+": "+str(MSE))
        y_pred = model.predict(X_test)
        mean_square_error = mean_squared_error(y_test, y_pred)
        mean_squared_errors.append(mean_square_error)

    mean_squared_errors = np.array(mean_squared_errors)
    mean = np.mean(mean_squared_errors)
    standard_deviation = np.std(mean_squared_errors)

    print('\n')
    print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors without normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
    print("Mean: "+str(mean))
    print("Standard Deviation: "+str(standard_deviation))
    return (mean, standard_deviation)

In [22]:
mean_a, std_a = iterate(50,50)

MSE 1: 853.0998396888906
MSE 2: 494.5173890937879
MSE 3: 283.1975157407495
MSE 4: 165.22060809089143
MSE 5: 143.523170495882
MSE 6: 123.23263080683341
MSE 7: 137.19905495257825
MSE 8: 105.43164867413468
MSE 9: 120.85986594783449
MSE 10: 110.83097720840603
MSE 11: 113.9254932094932
MSE 12: 101.44837724506662
MSE 13: 115.62173869308916
MSE 14: 117.08211603751074
MSE 15: 110.66500518700066
MSE 16: 107.12284515282097
MSE 17: 104.47870312539504
MSE 18: 115.40487098076582
MSE 19: 97.57155083221139
MSE 20: 118.87278890764057
MSE 21: 101.62798958688877
MSE 22: 105.40854676873167
MSE 23: 106.24594933469704
MSE 24: 108.32465653897874
MSE 25: 108.27689006027666
MSE 26: 102.9147535404341
MSE 27: 123.03434027514412
MSE 28: 108.68413360990753
MSE 29: 109.41963143950527
MSE 30: 131.23252048924516
MSE 31: 130.80620278664006
MSE 32: 119.88219117346705
MSE 33: 106.40082888310009
MSE 34: 108.49766550403582
MSE 35: 115.6200944227694
MSE 36: 125.48982270867307
MSE 37: 146.01680558090456
MSE 38: 113.7233052

<div id='partb'></div>

# Part-(B)

Result from previous part- 

The mean and standard deviation of 50 mean squared errors without normalized data. Total number of epochs for each training is: 50

Mean: 51.87652089428187
Standard Deviation: 8.24017812052958


**Here first job is normalisation.**

**--Then split data in train and test**

**--Build the modal**

**--Iterate for MSE and show mean and deviation**

In [19]:
#Normalising

predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [20]:
# Let's split the data in training and testing

X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)

In [21]:
mean_b, std_b = iterate(epochs=50,total_mean_squared_errors=50)

MSE 1: 305.3779777847833
MSE 2: 178.93250173426753
MSE 3: 132.32914620309973
MSE 4: 125.81570997515928
MSE 5: 121.70389127268375
MSE 6: 108.30381034184428
MSE 7: 136.13804527702456
MSE 8: 101.1601958043367
MSE 9: 121.21861222646768
MSE 10: 115.80331203621182
MSE 11: 102.15777005192531
MSE 12: 101.59565929918999
MSE 13: 114.90790459098939
MSE 14: 116.50257601938587
MSE 15: 109.6730386431549
MSE 16: 109.95747592765537
MSE 17: 106.85791351417122
MSE 18: 100.36113381000013
MSE 19: 94.91264276597106
MSE 20: 113.74643334447373
MSE 21: 97.43560312018039
MSE 22: 102.39871721977555
MSE 23: 108.82988714323075
MSE 24: 84.91982637177007
MSE 25: 67.18340124975902
MSE 26: 64.82489522298177
MSE 27: 63.546998403604746
MSE 28: 67.03255014512145
MSE 29: 67.26886012253252
MSE 30: 65.67281039092919
MSE 31: 67.31043164398292
MSE 32: 60.17278233006548
MSE 33: 63.784514245091906
MSE 34: 64.33909577304877
MSE 35: 64.4074663081987
MSE 36: 74.47281883758248
MSE 37: 66.92173555367972
MSE 38: 66.37862193777337
MS

**Here now we will see the change in mean and standard deviation**

In [27]:
change=[(mean_a, std_a),(mean_b, std_b)]
pd.DataFrame(change,index=['A','B'],columns=['Mean','STD'])

Mean       STD
A  40.884407  4.094926
B  39.704064  4.558720

<div id='partc'></div>


# Part-C

**Repeating Part B but for epochs=100 this time for training.**

In [ ]:
mean_c, std_c = iterate(epochs=100,total_mean_squared_errors=50)

In [ ]:
change=[(mean_a, std_a),(mean_b, std_b),(mean_c, std_c)]
pd.DataFrame(change,index=['A','B'],columns=['Mean','STD'])